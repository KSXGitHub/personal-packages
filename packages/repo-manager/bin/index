#! /usr/bin/env node
const process = require('process')
process.argv[0] = 'repo-manager'
require('./main').main().then(
  status => process.exit(status),
  error => {
    console.error(error)
    process.exit(-1)
  }
)
