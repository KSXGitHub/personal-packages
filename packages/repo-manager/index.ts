export * from './lib/types'
export * from './lib/constants'
export * from './lib/errors'
export * from './lib/parse'
export * from './lib/clone'
export * from './lib/pull-or-clone'
export * from './lib/create'
