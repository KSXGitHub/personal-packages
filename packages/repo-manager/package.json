{
  "name": "@khai96x/repo-manager",
  "version": "0.0.0",
  "description": "A way to manage my local git repositories",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/KSXGitHub/personal-packages#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/KSXGitHub/personal-packages.git"
  },
  "bugs": {
    "url": "https://github.com/KSXGitHub/personal-packages/issues"
  },
  "dependencies": {
    "@tsfun/pipe": "^0.0.0",
    "@tsfun/result": "^0.0.2",
    "isomorphic-git": "^0.60.1",
    "tslib": "^1.10.0",
    "@types/node": "^12.7.1"
  }
}
