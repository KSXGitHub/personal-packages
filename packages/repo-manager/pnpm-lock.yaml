dependencies:
  '@tsfun/pipe': 0.0.1
  '@tsfun/result': 0.0.3
  '@types/fs-extra': 8.0.0
  '@types/node': 12.7.2
  fs-extra: 8.1.0
  isomorphic-git: 0.64.0
  tslib: 1.10.0
lockfileVersion: 5.1
packages:
  /@tsfun/pipe/0.0.1:
    dependencies:
      '@tsfun/tap': 0.0.0
      '@types/node': 12.7.2
      tslib: 1.10.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-l1Rxz9zoR4taPtcRkaAA4nPlS8XoR/q6yRVYWqY2oXPNy64/OkxiwiRaCp6RSqrxzJPPzACRXXq3HOZUCh8wCw==
  /@tsfun/prv-option-result-common/0.0.1:
    dependencies:
      '@types/node': 11.13.19
      tslib: 1.10.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-FI4Zcc3I4lal4cMdBpWp0O86Ep4G/o8Q7cD47qLXaLa5kom+6pYRO4+XYBjSLKK+2h6rGWd1IvILmmMyAGRN3g==
  /@tsfun/result/0.0.3:
    dependencies:
      '@tsfun/prv-option-result-common': 0.0.1
      '@tsfun/tap': 0.0.0
      '@types/node': 12.7.2
      tslib: 1.10.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-L/+QX438aOALaFVtj3az61dhhtqMUjpallnmAlxWvF862uNToeTZdD9VMojM+8K7HOJvpUMswEp3uRTbKYtlAw==
  /@tsfun/tap/0.0.0:
    dependencies:
      '@types/node': 12.7.2
      tslib: 1.10.0
    dev: false
    resolution:
      integrity: sha512-6NPk8oXn4sXqxdy3bjXlcd+E9NQB+emivkT+O0zVc6PeakN9Vt2Bxczi5KoNVqQTLSuFUvDDLzSvbsYelJJpEQ==
  /@types/fs-extra/8.0.0:
    dependencies:
      '@types/node': 12.7.2
    dev: false
    resolution:
      integrity: sha512-bCtL5v9zdbQW86yexOlXWTEGvLNqWxMFyi7gQA7Gcthbezr2cPSOb8SkESVKA937QD5cIwOFLDFt0MQoXOEr9Q==
  /@types/node/11.13.19:
    dev: false
    resolution:
      integrity: sha512-tLRDU1hmcWamtgRT2iVRdraAQVGFQGgtcqracSo9XyMN1VeZLSVGb8RJJxVqab7UGbijoUijGPVFMjmqzyZIUw==
  /@types/node/12.7.2:
    dev: false
    resolution:
      integrity: sha512-dyYO+f6ihZEtNPDcWNR1fkoTDf3zAK3lAABDze3mz6POyIercH0lEUawUFXlG8xaQZmm1yEBON/4TsYv/laDYg==
  /async-lock/1.2.2:
    dev: false
    resolution:
      integrity: sha512-uczz62z2fMWOFbyo6rG4NlV2SdxugJT6sZA2QcfB1XaSjEiOh8CuOb/TttyMnYQCda6nkWecJe465tGQDPJiKw==
  /base64-js/0.0.2:
    dev: false
    engines:
      node: '>= 0.4'
    resolution:
      integrity: sha1-Ak8Pcq+iW3X5wO5zzU9V7Bvtl4Q=
  /bops/0.0.7:
    dependencies:
      base64-js: 0.0.2
      to-utf8: 0.0.1
    dev: false
    resolution:
      integrity: sha1-tKClqDmkBkVK8P4FqLkaenZqVOI=
  /clean-git-ref/2.0.1:
    dev: false
    resolution:
      integrity: sha512-bLSptAy2P0s6hU4PzuIMKmMJJSE6gLXGH1cntDu7bWJUksvuM+7ReOK61mozULErYvP6a15rnYl0zFDef+pyPw==
  /crc-32/1.2.0:
    dependencies:
      exit-on-epipe: 1.0.1
      printj: 1.1.2
    dev: false
    engines:
      node: '>=0.8'
    hasBin: true
    resolution:
      integrity: sha512-1uBwHxF+Y/4yF5G48fwnKq6QsIXheor3ZLPT80yGBV1oEUwpPojlEhQbWKVw1VwcTQyMGHK1/XMmTjmlsmTTGA==
  /decompress-response/3.3.0:
    dependencies:
      mimic-response: 1.0.1
    dev: false
    engines:
      node: '>=4'
    resolution:
      integrity: sha1-gKTdMjdIOEv6JICDYirt7Jgq3/M=
  /diff-lines/1.1.1:
    dependencies:
      diff: 3.5.0
    dev: false
    resolution:
      integrity: sha512-Oo5JzEEriF/+T0usOeRP5yOzr6SWvni2rrxvIgijMZSxPcEvf8JOvCO5GpnWwkte7fcOgnue/f5ECg1H9lMPCw==
  /diff/3.5.0:
    dev: false
    engines:
      node: '>=0.3.1'
    resolution:
      integrity: sha512-A46qtFgd+g7pDZinpnwiRJtxbC1hpgf0uzP3iG89scHk0AUC7A1TGxf5OiiOUv/JMZR8GOt8hL900hV0bOy5xA==
  /exit-on-epipe/1.0.1:
    dev: false
    engines:
      node: '>=0.8'
    resolution:
      integrity: sha512-h2z5mrROTxce56S+pnvAV890uu7ls7f1kEvVGJbw1OlFH3/mlJ5bkXu0KRyW94v37zzHPiUd55iLn3DA7TjWpw==
  /fs-extra/8.1.0:
    dependencies:
      graceful-fs: 4.2.2
      jsonfile: 4.0.0
      universalify: 0.1.2
    dev: false
    engines:
      node: '>=6 <7 || >=8'
    resolution:
      integrity: sha512-yhlQgA6mnOJUKOsRUFsgJdQCvkKhcz8tlZG5HBQfReYZy46OwLcY+Zia0mtdHsOo9y/hP+CxMN0TU9QxoOtG4g==
  /git-apply-delta/0.0.7:
    dependencies:
      bops: 0.0.7
      varint: 0.0.3
    dev: false
    resolution:
      integrity: sha1-+3auFEVA15RAtSsx3gPmPJk8chk=
  /globalyzer/0.1.4:
    dev: false
    resolution:
      integrity: sha512-LeguVWaxgHN0MNbWC6YljNMzHkrCny9fzjmEUdnF1kQ7wATFD1RHFRqA1qxaX2tgxGENlcxjOflopBwj3YZiXA==
  /globrex/0.1.2:
    dev: false
    resolution:
      integrity: sha512-uHJgbwAMwNFf5mLst7IWLNg14x1CkeqglJb/K3doi4dw6q2IvAAmM/Y81kevy83wP+Sst+nutFTYOGg3d1lsxg==
  /graceful-fs/4.2.2:
    dev: false
    resolution:
      integrity: sha512-IItsdsea19BoLC7ELy13q1iJFNmd7ofZH5+X/pJr90/nRoPEX0DJo1dHDbgtYWOhJhcCgMDTOw84RZ72q6lB+Q==
  /ignore/5.1.4:
    dev: false
    engines:
      node: '>= 4'
    resolution:
      integrity: sha512-MzbUSahkTW1u7JpKKjY7LCARd1fU5W2rLdxlM4kdkayuCwZImjkpluF9CM1aLewYJguPDqewLam18Y6AU69A8A==
  /inherits/2.0.4:
    dev: false
    resolution:
      integrity: sha512-k/vGaX4/Yla3WzyMCvTQOXYeIHvqOKtnqBduzTHpzpQZzAskKMhZ2K+EnBiSM9zGSoIFeMpXKxa4dYeZIQqewQ==
  /isomorphic-git/0.64.0:
    dependencies:
      async-lock: 1.2.2
      clean-git-ref: 2.0.1
      crc-32: 1.2.0
      diff-lines: 1.1.1
      git-apply-delta: 0.0.7
      globalyzer: 0.1.4
      globrex: 0.1.2
      ignore: 5.1.4
      marky: 1.2.1
      minimisted: 2.0.0
      pako: 1.0.10
      pify: 4.0.1
      readable-stream: 3.4.0
      sha.js: 2.4.11
      simple-get: 3.0.3
    dev: false
    engines:
      node: '>=7.6.0'
    hasBin: true
    resolution:
      integrity: sha512-6zUjD3B5Y9kcK8s/DHBFcjH30Yt0Ewxj3YPzDUXqdmSz1BnWnrOhgKeOA5x+FB5dcqFPe2fLLeyrEpKAxIjDOQ==
  /jsonfile/4.0.0:
    dev: false
    optionalDependencies:
      graceful-fs: 4.2.2
    resolution:
      integrity: sha1-h3Gq4HmbZAdrdmQPygWPnBDjPss=
  /marky/1.2.1:
    dev: false
    resolution:
      integrity: sha512-md9k+Gxa3qLH6sUKpeC2CNkJK/Ld+bEz5X96nYwloqphQE0CKCVEKco/6jxEZixinqNdz5RFi/KaCyfbMDMAXQ==
  /mimic-response/1.0.1:
    dev: false
    engines:
      node: '>=4'
    resolution:
      integrity: sha512-j5EctnkH7amfV/q5Hgmoal1g2QHFJRraOtmx0JpIqkxhBhI/lJSl1nMpQ45hVarwNETOoWEimndZ4QK0RHxuxQ==
  /minimist/1.2.0:
    dev: false
    resolution:
      integrity: sha1-o1AIsg9BOD7sH7kU9M1d95omQoQ=
  /minimisted/2.0.0:
    dependencies:
      minimist: 1.2.0
    dev: false
    resolution:
      integrity: sha512-oP88Dw3LK/pdrKyMdlbmg3W50969UNr4ctISzJfPl+YPYHTAOrS+dihXnsgRNKSRIzDsrnV3eE2CCVlZbpOKdQ==
  /once/1.4.0:
    dependencies:
      wrappy: 1.0.2
    dev: false
    resolution:
      integrity: sha1-WDsap3WWHUsROsF9nFC6753Xa9E=
  /pako/1.0.10:
    dev: false
    resolution:
      integrity: sha512-0DTvPVU3ed8+HNXOu5Bs+o//Mbdj9VNQMUOe9oKCwh8l0GNwpTDMKCWbRjgtD291AWnkAgkqA/LOnQS8AmS1tw==
  /pify/4.0.1:
    dev: false
    engines:
      node: '>=6'
    resolution:
      integrity: sha512-uB80kBFb/tfd68bVleG9T5GGsGPjJrLAUpR5PZIrhBnIaRTQRjqdJSsIKkOP6OAIFbj7GOrcudc5pNjZ+geV2g==
  /printj/1.1.2:
    dev: false
    engines:
      node: '>=0.8'
    hasBin: true
    resolution:
      integrity: sha512-zA2SmoLaxZyArQTOPj5LXecR+RagfPSU5Kw1qP+jkWeNlrq+eJZyY2oS68SU1Z/7/myXM4lo9716laOFAVStCQ==
  /readable-stream/3.4.0:
    dependencies:
      inherits: 2.0.4
      string_decoder: 1.3.0
      util-deprecate: 1.0.2
    dev: false
    engines:
      node: '>= 6'
    resolution:
      integrity: sha512-jItXPLmrSR8jmTRmRWJXCnGJsfy85mB3Wd/uINMXA65yrnFo0cPClFIUWzo2najVNSl+mx7/4W8ttlLWJe99pQ==
  /safe-buffer/5.2.0:
    dev: false
    resolution:
      integrity: sha512-fZEwUGbVl7kouZs1jCdMLdt95hdIv0ZeHg6L7qPeciMZhZ+/gdesW4wgTARkrFWEpspjEATAzUGPG8N2jJiwbg==
  /sha.js/2.4.11:
    dependencies:
      inherits: 2.0.4
      safe-buffer: 5.2.0
    dev: false
    hasBin: true
    resolution:
      integrity: sha512-QMEp5B7cftE7APOjk5Y6xgrbWu+WkLVQwk8JNjZ8nKRciZaByEW6MubieAiToS7+dwvrjGhH8jRXz3MVd0AYqQ==
  /simple-concat/1.0.0:
    dev: false
    resolution:
      integrity: sha1-c0TLuLbib7J9ZrL8hvn21Zl1IcY=
  /simple-get/3.0.3:
    dependencies:
      decompress-response: 3.3.0
      once: 1.4.0
      simple-concat: 1.0.0
    dev: false
    resolution:
      integrity: sha512-Wvre/Jq5vgoz31Z9stYWPLn0PqRqmBDpFSdypAnHu5AvRVCYPRYGnvryNLiXu8GOBNDH82J2FRHUGMjjHUpXFw==
  /string_decoder/1.3.0:
    dependencies:
      safe-buffer: 5.2.0
    dev: false
    resolution:
      integrity: sha512-hkRX8U1WjJFd8LsDJ2yQ/wWWxaopEsABU1XfkM8A+j0+85JAGppt16cr1Whg6KIbb4okU6Mql6BOj+uup/wKeA==
  /to-utf8/0.0.1:
    dev: false
    resolution:
      integrity: sha1-0Xrqcv8vujm55DYBvns/9y4ImFI=
  /tslib/1.10.0:
    dev: false
    resolution:
      integrity: sha512-qOebF53frne81cf0S9B41ByenJ3/IuH8yJKngAX35CmiZySA0khhkovshKK+jGCaMnVomla7gVlIcc3EvKPbTQ==
  /universalify/0.1.2:
    dev: false
    engines:
      node: '>= 4.0.0'
    resolution:
      integrity: sha512-rBJeI5CXAlmy1pV+617WB9J63U6XcazHHF2f2dbJix4XzpUF0RS3Zbj0FGIOCAva5P/d/GBOYaACQ1w+0azUkg==
  /util-deprecate/1.0.2:
    dev: false
    resolution:
      integrity: sha1-RQ1Nyfpw3nMnYvvS1KKJgUGaDM8=
  /varint/0.0.3:
    dev: false
    resolution:
      integrity: sha1-uCHemwSzizzSL3LBjZSp+3KrNRg=
  /wrappy/1.0.2:
    dev: false
    resolution:
      integrity: sha1-tSQ9jz7BqjXxNkYFvA0QNuMKtp8=
specifiers:
  '@tsfun/pipe': ^0.0.1
  '@tsfun/result': ^0.0.3
  '@types/fs-extra': ^8.0.0
  '@types/node': ^12.7.2
  fs-extra: ^8.1.0
  isomorphic-git: ^0.64.0
  tslib: ^1.10.0
