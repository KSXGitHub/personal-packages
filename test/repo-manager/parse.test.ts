import { ok, err, unwrap } from '@tsfun/result'
import { parseGitUrl, createGitUrl, GitUrlInfo, Host } from '@khai96x/repo-manager'

describe('parseGitURL', () => {
  describe('valid links', () => {
    describe('https://github.com/org/repo.git', () => {
      const get = () => parseGitUrl('https://github.com/org/repo.git')

      it('returns expected result', () => {
        expect(get()).toEqual(ok<GitUrlInfo>({
          host: Host.GitHub,
          owner: 'org',
          name: 'repo'
        }))
      })

      it('matches snapshot', () => {
        expect(get()).toMatchSnapshot()
      })
    })

    describe('https://github.com/org/repo', () => {
      const get = () => parseGitUrl('https://github.com/org/repo')

      it('returns expected result', () => {
        expect(get()).toEqual(ok<GitUrlInfo>({
          host: Host.GitHub,
          owner: 'org',
          name: 'repo'
        }))
      })

      it('matches snapshot', () => {
        expect(get()).toMatchSnapshot()
      })
    })

    describe('https://gitlab.com/org/repo.git', () => {
      const get = () => parseGitUrl('https://gitlab.com/org/repo.git')

      it('returns expected result', () => {
        expect(get()).toEqual(ok<GitUrlInfo>({
          host: Host.GitLab,
          owner: 'org',
          name: 'repo'
        }))
      })

      it('matches snapshot', () => {
        expect(get()).toMatchSnapshot()
      })
    })

    describe('https://gitlab.com/org/repo', () => {
      const get = () => parseGitUrl('https://gitlab.com/org/repo')

      it('returns expected result', () => {
        expect(get()).toEqual(ok<GitUrlInfo>({
          host: Host.GitLab,
          owner: 'org',
          name: 'repo'
        }))
      })

      it('matches snapshot', () => {
        expect(get()).toMatchSnapshot()
      })
    })

    it('supports http', () => {
      expect(
        parseGitUrl('http://github.com/org/repo.git')
      ).toEqual(
        ok(expect.anything())
      )
    })
  })

  describe('invalid links', () => {
    describe('unknown protocol', () => {
      const get = () => parseGitUrl('unknown://github.com/org/repo.git')

      it('returns an err', () => {
        expect(get()).toEqual(err(expect.any(Error)))
      })

      it('contains expected properties', () => {
        expect(get()).toEqual(err(expect.objectContaining({
          protocol: 'unknown:'
        })))
      })

      it('matches snapshot', () => {
        expect(get()).toMatchSnapshot()
      })
    })

    describe('unknown host', () => {
      const get = () => parseGitUrl('https://unknown.com/org/repo.git')

      it('returns an err', () => {
        expect(get()).toEqual(err(expect.any(Error)))
      })

      it('contains expected properties', () => {
        expect(get()).toEqual(err(expect.objectContaining({
          hostname: 'unknown.com'
        })))
      })

      it('matches snapshot', () => {
        expect(get()).toMatchSnapshot()
      })
    })

    describe('excessive path', () => {
      const get = () => parseGitUrl('https://github.com/org/repo/tree/master/README.md')

      it('returns an err', () => {
        expect(get()).toEqual(err(expect.any(Error)))
      })

      it('contains expected properties', () => {
        expect(get()).toEqual(err(expect.objectContaining({
          excess: ['tree', 'master', 'README.md']
        })))
      })

      it('matches snapshot', () => {
        expect(get()).toMatchSnapshot()
      })
    })
  })
})

describe('createGitUrl', () => {
  it('host = GitHub', () => {
    expect(createGitUrl({
      host: Host.GitHub,
      owner: 'org',
      name: 'repo'
    })).toBe('https://github.com/org/repo.git')
  })

  it('host = GitLab', () => {
    expect(createGitUrl({
      host: Host.GitLab,
      owner: 'org',
      name: 'repo'
    })).toBe('https://gitlab.com/org/repo.git')
  })
})

describe('encode - decode', () => {
  describe('invalid characters', () => {
    const info: GitUrlInfo = {
      host: Host.GitHub,
      owner: 'abc/def ghi',
      name: 'foo bar/baz'
    }

    const url = 'https://github.com/abc%2Fdef%20ghi/foo%20bar%2Fbaz.git'

    it('info → createGitUrl', () => {
      expect(createGitUrl(info)).toBe(url)
    })

    it('info → createGitUrl → parseGitUrl', () => {
      expect(unwrap(parseGitUrl(createGitUrl(info)))).toEqual(info)
    })

    it('url → parseGitUrl', () => {
      expect(unwrap(parseGitUrl(url))).toEqual(info)
    })

    it('url → parseGitUrl → createGitUrl', () => {
      expect(createGitUrl(unwrap(parseGitUrl(url)))).toBe(url)
    })
  })

  describe('valid characters', () => {
    const info: GitUrlInfo = {
      host: Host.GitHub,
      owner: 'org',
      name: 'repo'
    }

    const url = 'https://github.com/org/repo.git'

    it('info → createGitUrl', () => {
      expect(createGitUrl(info)).toBe(url)
    })

    it('info → createGitUrl → parseGitUrl', () => {
      expect(unwrap(parseGitUrl(createGitUrl(info)))).toEqual(info)
    })

    it('url → parseGitUrl', () => {
      expect(unwrap(parseGitUrl(url))).toEqual(info)
    })

    it('url → parseGitUrl → createGitUrl', () => {
      expect(createGitUrl(unwrap(parseGitUrl(url)))).toBe(url)
    })
  })
})
