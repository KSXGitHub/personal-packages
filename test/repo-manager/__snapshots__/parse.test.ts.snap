// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseGitURL invalid links excessive path matches snapshot 1`] = `
Object {
  "error": [ExcessivePath: Path too long: /org/repo/tree/master/README.md],
  "tag": false,
}
`;

exports[`parseGitURL invalid links unknown host matches snapshot 1`] = `
Object {
  "error": [UnsupportedHostName: Hostname "unknown.com" is not supported],
  "tag": false,
}
`;

exports[`parseGitURL invalid links unknown protocol matches snapshot 1`] = `
Object {
  "error": [UnsupportedProtocol: Protocol "unknown:" is not supported],
  "tag": false,
}
`;

exports[`parseGitURL valid links https://github.com/org/repo matches snapshot 1`] = `
Object {
  "tag": true,
  "value": GitUrlInfoInstance {
    "host": "github",
    "name": "repo",
    "owner": "org",
  },
}
`;

exports[`parseGitURL valid links https://github.com/org/repo.git matches snapshot 1`] = `
Object {
  "tag": true,
  "value": GitUrlInfoInstance {
    "host": "github",
    "name": "repo",
    "owner": "org",
  },
}
`;

exports[`parseGitURL valid links https://gitlab.com/org/repo matches snapshot 1`] = `
Object {
  "tag": true,
  "value": GitUrlInfoInstance {
    "host": "gitlab",
    "name": "repo",
    "owner": "org",
  },
}
`;

exports[`parseGitURL valid links https://gitlab.com/org/repo.git matches snapshot 1`] = `
Object {
  "tag": true,
  "value": GitUrlInfoInstance {
    "host": "gitlab",
    "name": "repo",
    "owner": "org",
  },
}
`;
